### PR TITLE
Update key rotation guidelines for encryption at rest

### LINKED ISSUE #39477: Security information about AES-GCM key rotation are misleading
The official documentation for encryption at rest providers has some pretty confusing security guidelines for AES-GCM: https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/#providers.

I am no security expert but the mention of having to rotate an AES-GCM key every 200k writes is pretty confusing to me and I couldn't find the origin of this value anywhere. In k/k we are using AES-GCM with the default random nonce of 96 bits provided by go/crypto which should allow for way more key use than the theoretical 200k from the doc.

If we rely on the guidelines from the NIST in https://tsapps.nist.gov/publication/get_pdf.cfm?pub_id=51288#page=29, we should theoretically be able to use an AES-GCM key up to 2^32 times with a very low risk of having a nonce collision. Even in the event where there is a collision, the probability that the encrypted objects with the same nonce are present at the same time in etcd is very small making the threat level very low.

/cc @enj @ibihim @tkashem

### PR DESCRIPTION
**Pull request by dgrisonnet**
Add general guidelines for keys rotations and update the write-based requirements for `aesgcm` from `200 000` to `2^32`. In addition to that, the documentation now includes a complete explanation of the logic behind these numbers.

Fixes #39477

### CODE REVIEW

**Comment by sftim** [comment]
Hi. Please rebase this PR against _main_ - there was a breaking change to our CVE feed.

File: content/en/docs/tasks/administer-cluster/encrypt-data.md [hunk 1]
```diff
@@ -321,6 +321,57 @@ Then run the following command to force decrypt all Secrets:
...
+
+$$p(k) \approx 1 - e^{-{\frac{(k^2)}{2^{97}}}}$$
+
+A negligeable chance of collision by current security standards is $1/2^{32}$ so in Kubernetes' case we want to make sure that the chance of a none collision is less than one in $2^{32}$
+
+$$p(k) \approx 1 - e^{-{\frac{(k^2)}{2^{97}}}} \leq 2^{32}$$
+
+The smallest power of 2 for which this equation is satified is $2^{32}$
+
+$$1-e^{-\frac{2^{32}}{2^{97}}} \approx 1.164\times{10}^{-10} \leq \frac{1}{2^{32}} \approx 2.32\times{10}^{-10} \leq 1-e^{-\frac{2^{33}}{2^{97}}} \approx 4.656\times{10}^{-10}$$
```
Comment (dgrisonnet) [inline content/en/docs/tasks/administer-cluster/encrypt-data.md:371]: I tried to solve the formula with `k` as the unknown and find its value based on `p(k) <= 2^32`, but couldn't get the result I wanted. For the record, the logic to solve the birthday paradox with that kind of equation is explained in https://en.wikipedia.org/wiki/Birthday_problem under `An upper bound on the probability and a lower bound on the number of people`.

**Comment by dgrisonnet** [comment]
@sftim the math formulas are broken in the preview, do you know how I can fix them?

https://deploy-preview-39747--kubernetes-io-main-staging.netlify.app/docs/tasks/administer-cluster/encrypt-data/

**Comment by sftim** [comment]
> @sftim the math formulas are broken in the preview, do you know how I can fix them?
> 
> https://deploy-preview-39747--kubernetes-io-main-staging.netlify.app/docs/tasks/administer-cluster/encrypt-data/

We don't support maths formulas. Consider filing a feature request (label it **area/web-development**)? I don't think enabling these should be hard, as upstream Docsy does have support.

**Comment by dgrisonnet** [comment]
Could you point me to the part of the code I need to modify to support that? I might as well add the feature, while I am at it

**Comment by sftim** [comment]
> Could you point me to the part of the code I need to modify to support that? I might as well add the feature, while I am at it

I'm afraid I haven't looked. Try https://www.google.com/search?q="docsy"+math+support

**Comment by dgrisonnet** [comment]
I meant on the k/website side. But I can try to find my way in the repo.

**Comment by dgrisonnet** [comment]
Found, I will update the PR, thanks for pointing out to docsy

**Comment by dgrisonnet** [comment]
@sftim adding support for scientific formulas is a bit more complex than what I had anticipated.

I originally tried enabling KaTeX in config.toml following this docsy documentation: https://www.docsy.dev/docs/adding-content/diagrams-and-formulae/, but for some reasons it didn't work for Kubernetes. I tried on a brand new hugo website and following the steps from the doc worked as expected, so it seems to be specific to Kubernetes.

The first thing I noticed is that the website is using an old version of docsy and KaTeX format as changed a bit since then. Here the correct doc for what we are trying to achieve: https://github.com/google/docsy/blob/v0.2.0/userguide/content/en/docs/Adding%20content/diagrams-and-formulae/index.md.

But even when following this v0.2.0 doc, the formulas weren't rendered properly and there wasn't any logs/info in the web console that I could help my investigation.

While looking into that, I also noticed that Kubernetes website supports `mermaid` and since the docsy configuration is very similar, I thought I would've a look there to help me.
To my surprise mermaid is not currently enabled via docsy but rather there is a custom shortocode for it: https://github.com/kubernetes/website/blob/main/layouts/shortcodes/mermaid.html.

So I tried my chance at configuring mermaid via docsy to see if my problem was specific to katex. However after following the doc and enabling mermaid in config.toml. I was greeted by a webpage not rendereding the diagram and the following error in the console:
```js
Uncaught ReferenceError: mermaid is not defined
    at main.js:197:25
    at main.js:200:3
``` 

```js
197: var settings = norm(mermaid.mermaidAPI.defaultConfig, params);
```

From that observation, my current theory is that enabling docsy plugins such as katex and mermaid in the main config.toml doesn't work in Kubernetes' website special case because the pieces of JS that should be imported in the web pages are not. Do you perhaps have any idea what could cause a conflict that would prevent importing these libraries? 

As a temporary solution, would it perhaps make sense to create a new shortcode for katex similarly to what we are already doing for mermaid?

**Comment by sftim** [comment]
The Mermaid bug is https://github.com/kubernetes/website/issues/31960

Perhaps we can omit the formulas for this PR and open another PR to add those details as a follow-up?

**Comment by dgrisonnet** [comment]
Not yet since I wanted to solve it myself but it turns out to be a bit too complex. I'll do that once I've updated this PR.

**Comment by dgrisonnet** [comment]
I fixed the rendering of the scientific formulas. The latest version of the doc can be found at https://deploy-preview-39747--kubernetes-io-main-staging.netlify.app/docs/tasks/administer-cluster/encrypt-data/.

@sftim @enj @ibihim can you have a look please.

**Comment by sftim** [comment]
@dgrisonnet we'd like to have these changes, if you've got the capacity to update given the v1.27 release

**Comment by aramase** [comment]
/assign @enj @smarterclayton 

could you review this when you get a chance?

**Comment by kbhawkey** [comment]
@dgrisonnet ,
Can you rebase your pull request or can I close it?
Thanks

**Comment by divya-mohan0209** [comment]
Hello @dgrisonnet , we'd like to see this change merged. Would you have the capacity to work on this PR? Please let us know by 25th August, 2023, failing which we shall be closing the PR.

**Comment by dgrisonnet** [comment]
Hi @divya-mohan0209, thank you for the reminder, I'll try to work on it again in the upcoming weeks.

**Comment by divya-mohan0209** [comment]
Thank you for confirming @dgrisonnet

**Comment by reylejano** [comment]
Hi @dgrisonnet , do you have some time to rebase this PR as there is a merge conflict.
If there is no rebase, a PR wrangler may close this PR

**Comment by kbhawkey** [comment]
Hi @dgrisonnet .
I plan to close this pull request. When you have time to rebase, feel free to reopen the PR.
Thanks
/close

**Pull request closed without merging**